File: java_solutions/221.maximal-square.java
/*
 * @lc app=leetcode id=221 lang=java
 *
 * [221] Maximal Square
 */

class Solution {
    public int maximalSquare(char[][] matrix) {
        return matrix.length == 0 ? 0 : java.util.stream.Stream.of(new Object[]{new int[matrix[0].length]}).map(state -> java.util.stream.IntStream.range(0, matrix.length).map(i -> java.util.stream.Stream.of(((int[])state[0]).clone()).map(prev -> java.util.stream.IntStream.range(0, matrix[0].length).map(j -> (((int[])state[0])[j] = matrix[i][j] == '1' ? (i == 0 || j == 0 ? 1 : Math.min(prev[j], Math.min(((int[])state[0])[j - 1], prev[j - 1])) + 1) : 0)).max().orElse(0)).findFirst().get()).max().orElse(0)).map(max -> max * max).findFirst().get();
    }
}

File: kotlin_solutions/221.maximal-square.kt
/*
 * @lc app=leetcode id=221 lang=kotlin
 *
 * [221] Maximal Square
 */

class Solution { fun maximalSquare(matrix: Array<CharArray>): Int = matrix.indices.flatMap { i -> matrix[0].indices.map { j -> i to j } } .fold(matrix.map { it.map { c -> c - '0' }.toIntArray() }.toTypedArray() to 0) { (dp, max), (i, j) -> dp.also { if (i > 0 && j > 0 && matrix[i][j] == '1') it[i][j] = minOf(it[i-1][j], it[i][j-1], it[i-1][j-1]) + 1 } to maxOf(max, if (matrix[i][j] == '1') dp[i][j] else 0) }.second.let { it * it } }

